File: src/algs4/1.3.37/Josephus.java
public class Josephus {
    private Person current;
    private Person last;
    private int N;

    private class Person {
        int pos;
        Person next;
    }

    public Josephus(int size) {
        N = 0;

        // sit in a circle
        for (int i = 0; i < size; i++) {
            Person one = new Person();
            one.pos = i;
            if (i == 0) {
                current = one;
            }
            if (last == null) {
                last = one;
            } else {
                last.next = one;
                last = one;
            }
            if (i == size - 1) {
                one.next = current;
            }
            N++;
        }
    }

    public Person killAfter(Person target) {
        if (N == 2) {
            Person toKill = target.next;
            target.next = null;
            N--;
            System.out.println("Kill " + toKill.pos);
            return toKill;
        } else if (N > 2) {
            Person toKill = target.next;
            target.next = target.next.next;
            System.out.println("Kill " + toKill.pos);
            N--;
            return toKill;
        } else {
            System.out.println("No man to kill");
            return null;
        }
    }

    public int getLuckyOne(int call) {

        while (N > 1) {
            System.out.println("New round begin, current person position: " + current.pos);
            Person personToKill = current;
            for (int i = 1; i <= call - 2; i++) {
                personToKill = personToKill.next;
            }
            Person toKill = killAfter(personToKill);
            if (N > 2) {
                current = personToKill.next;
            } else {
                current = toKill.next;
            }
            System.out.println("Round close, " + N + " person left");
        }
        System.out.println("All rounds finished " + current.pos);
        return current.pos;
    }

    public static void main(String[] args) {
        Josephus j = new Josephus(7);
        System.out.println("Safe position: " + j.getLuckyOne(3));
    }
}

File: playground/josephus.js
/**
 * @param  {number} n Total number of people
 * @param  {number} k Person to kill
 * @return {number}
 */
function safe(n, k, toSave = 1) {
  let i = 1;
  let start;
  let last = null;
  while (i <= n) {
    let node = new Node(i);
    if (i === 1) start = node;
    if (last !== null) last.next = node;
    last = node;
    i++;
  }
  last.next = start;

  let current = start;
  while (i > toSave) {
    let j = 1;
    while (j < k) {
      last = current;
      current = current.next;
      j++;
    }
    let toKill = current;
    let next = toKill.next;
    last.next = next;
    current = next;
    i--;
  }

  return current.pos;
}

function Node(pos) {
  this.pos = pos;
  this.next = null;
}

var expect = require('chai').expect;

expect(safe(3, 2)).to.equal(3);
expect(safe(5, 2)).to.equal(3);
expect(safe(6, 3)).to.equal(1);